util/mount.fuse.c

```diff
@@ -80,6 +93,116 @@ static char *add_option(const char *opt, char *options)
 	return options;
 }
 
+static int prepare_fuse_fd(const char *mountpoint, const char* subtype,
+			   const char *options)
+{
+	int subtype_len = strlen(subtype) + 9;
+	char* options_copy = xrealloc(NULL, subtype_len);
+	snprintf(options_copy, subtype_len, "subtype=%s", subtype);
+	options_copy = add_option(options, options_copy);
+
+	const char *argv[] = { progname, "-o", options_copy };
+	struct fuse_args args = FUSE_ARGS_INIT(sizeof(argv) / sizeof(argv[0]),
+					       (char**) argv);
+	struct fuse_lowlevel_ops dummy_ops;
+	memset(&dummy_ops, 0, sizeof(dummy_ops));
+	struct fuse_session *session = fuse_session_new(
+			&args, &dummy_ops, sizeof(dummy_ops), NULL);
+	free(options_copy);
+	if (session == NULL || fuse_session_mount(session, mountpoint) == -1) {
+		exit(1);
+	}
+
+	/*
+	 * Duplicate the FUSE file descriptor to obtain a copy that's still
+	 * valid after fuse_session_destroy closes its copy. Note that this
+	 * conveniently also gives us a file descriptor with the CLOEXEC flag
+	 * clear, which is required to pass it across exec.
+	 */
+	int fuse_fd = fuse_session_fd(session);
+	fuse_fd = dup(fuse_fd);
+	if (fuse_fd == -1) {
+		fprintf(stderr, "%s: Failed to duplicate FUSE fd: %s\n",
+			progname, strerror(errno));
+		exit(1);
+	}
+	fuse_session_destroy(session);
+
+	return fuse_fd;
+}
+
+#ifdef linux
+static void set_capabilities(uint64_t caps)
+{
+	/*
+	 * This invokes the capset syscall directly to avoid the libcap
+	 * dependency, which isn't really justified just for this.
+	 */
+	struct __user_cap_header_struct header = {
+		.version = _LINUX_CAPABILITY_VERSION_3,
+		.pid = 0,
+	};
+	struct __user_cap_data_struct data[2];
+	memset(data, 0, sizeof(data));
+	data[0].effective = data[0].permitted = caps;
+	data[1].effective = data[1].permitted = caps >> 32;
+	if (syscall(SYS_capset, &header, data) == -1) {
+		fprintf(stderr, "%s: Failed to drop capabilities: %s\n",
+			progname, strerror(errno));
+		exit(1);
+	}
+}
+
+static void drop_privileges(void)
```

Comment:
This will only work if launched from something where the bits are not already locked -- not sure if that's a problem.

Reply:
That, and we need CAP_SETPCAP. This is all behind an option though, so people who do their own sandboxing aren't forced to use it. There might be a point to offer an option that just enables pass_fuse_fd, but doesn't attempt to drop privileges... I guess this can be added once there's demand for it?